### ISSUE TITLE
Code Completion Not Working for Sub-classes or Other Files

### ISSUE THREAD

**Issue by SolarLune** [godotengine/godot#1039]
OK, so it seems like code completion only works for built-in classes and types. If you define a sub-class in a GDScript file, you can't access that class's members or functions. Also, if you define anything in a GDScript file itself (using the file itself as the "class"), you won't be able to get any completion on its attributes. An example would be

```
const STATE_RELEASED = 3
```

in `input.gd`, and trying to use

```
var input = preload("res://input.gd").new()
```

to refer to the file. Printing out `input.STATE_RELEASED` works, but the completion displays nothing from the `input.gd` file.

**Comment by SolarLune** [comment]
This issue seems to be slightly resolved somewhat. I'm guessing the .new() was unnecessary in my example script? Seems like
```
var input = preload("res://input.gd")

input.STATE_RELEASED
```

completes correctly. However, the instantiated versions of classes still don't complete correctly, regardless of whether it was a class file (`preload("res://custom class file.gd").new()`) or a class instance (`v = local_custom_class.new()`).

As a side-note, the preload() function itself isn't auto-completing.

**Comment by SolarLune** [comment]
Appending on an additional found case - if you extend a custom .gd file from a string, code completion won't work. At least in my case, the .gd file is in another directory, so it's `extends "../gameobject.gd"`, in case that factors into the bug at all.

**Comment by okamstudio** [comment]
yeah i'm just realizing that completing instantiated classes is more
difficult than it seems, need to add to the internal type representation
which script it uses also

On Tue, Jan 6, 2015 at 12:55 PM, Sol <notifications@github.com> wrote:

> Appending on an additional found case - if you extend a custom .gd file
> from a string, code completion won't work. At least in my case, the .gd
> file is in another directory, so it's extends "../gameobject.gd", in case
> that factors into the bug at all.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/okamstudio/godot/issues/1039#issuecomment-68883723>.
>



-- 

*OkamStudio*

**Comment by vnen** [comment]
Just as a note: code completion also doesn't work for instanced scenes.

**Comment by akien-mga** [comment]
Not critical for the upcoming 2.1, so moving to the next milestone.

**Comment by akien-mga** [comment]
Partly fixed by #8062, but apparently not fully.

**Comment by akien-mga** [comment]
First of all thank you for your report and sorry for the delay.

We released Godot 3.0 in January 2018 after 18 months of work, fixing many old issues either directly, or by obsoleting/replacing the features they were referring to.

We still have hundreds of issues whose relevance/reproducibility needs to be checked against the [current stable version](https://godotengine.org/download), and that's where you can help us.
**Could you check if the issue that you described initially is still relevant/reproducible in Godot 3.0 or any newer version, and comment about it here?**

For bug reports, please also make sure that the issue contains detailed steps to reproduce the bug and, if possible, a zipped project that can be used to reproduce it right away. This greatly speeds up debugging and bugfixing tasks for our contributors.

Our Bugsquad will review this issue more in-depth in 15 days, and potentially close it if its relevance could not be confirmed.

Thanks in advance.

*Note: This message is being copy-pasted to many "stale" issues (90+ days without activity). It might happen that it is not meaningful for this specific issue or appears oblivious of the issue's context, if so please comment to notify the Bugsquad about it.*
